### PR TITLE
Update sdcard etc

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description

This PR fixes a number of issues related to the SDCARD class in System.IO.Filesystem

- The mapping of gpio pins for S3 and P4 targets.
- Allowing the card detect to be active high or low
- Adds some missing unit tests for a previous fix for list files with directories in same folder.
- Card detct has been broken for a while. Fixes a problem where interrupts for card detect was not working due the same event names being used for card detect and when drives are mounted.  
- Added an Auto mount on card detect (optional)
- Added feature to allow multiple cards (which was already existing in firmware). There is now a slotIndex parameter which denotes which card slot is being worked with. ESP32 allow 2 x slots for SDIO pins and we can have additional slots on SPI bus.
- Separate the Card detect parameter in to separate parameter class.
- Due to all these changes the API has been changed and not backward compatible.

This assembly requires updated NanoClr
- Includes native code fix for unreliable GPIO event when debounce was enabled.
- SDcard pins mapping
- Update events 
 
Also updated Hardware.ESP32
- New pin mapping option for SDMMC1 and SDMMC2 devices.

## Motivation and Context

- Fixes nanoFramework/Home#1447
- Fixes nanoFramework/Home#1475

## How Has This Been Tested?

Tested locally with updated mount sample

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [] New feature (non-breaking change which adds functionality to code)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

## Summary by CodeRabbit

- **New Features**
	- Updated the `Nerdbank.GitVersioning` package to a newer version, which may include improved performance, bug fixes, or new functionalities for users.
  
- **Bug Fixes**
	- This update is expected to address previously known issues related to the earlier version of the package.

### CODE REVIEW

**Review by josesimoes — COMMENTED** [review]
There is already `RemovableDriveEventArgs` why should it be replaced with `CardDetectChangedEventArgs`? I get it that there is this new `slotIndex`, but that can be added to the existing class.
Moreover, card implies that it's _card_ only opposed to removable driver that works for cards and USB MSD (any removable media).

File: System.IO.FileSystem/System.IO.FileSystem.nfproj [hunk 1]
```diff
@@ -78,19 +82,19 @@
     <Content Include="packages.lock.json" />
   </ItemGroup>
   <ItemGroup>
-    <Reference Include="mscorlib, Version=1.15.6.0, Culture=neutral, PublicKeyToken=c07d481e9758c731">
+    <Reference Include="mscorlib">
```
Comment (josesimoes) [inline System.IO.FileSystem/System.IO.FileSystem.nfproj:85]: These versions and sing key tokens need to be here.

File: System.IO.FileSystem/nanoFramework/SDCardCDParameters.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// Copyright (c) .NET Foundation and Contributors
+// See LICENSE file in the project root for full license information.
+//
+
+namespace nanoFramework.System.IO.FileSystem
+{
+    /// <summary>
+    /// Parameter used for Card detection when creating a SDcard instance.
+    /// </summary>
+    public class SDCardCDParameters
```
Comment (josesimoes) [inline System.IO.FileSystem/nanoFramework/SDCardCDParameters.cs:11]: The name could become large, still its preferable than using initials to condense the action that is, in fact, the reason for this class to exist: Card Detection. I suggest `CardDetectionParameters`.

**Review by josesimoes — COMMENTED** [review]
@AdrianSoundy I think that with these changes in `StorageEventManager` you are recreating the logic that already exists at the native end... 
There is an API to mount and unmount that calls the appropriate driver handler. 
There is already a volume list that takes care of storing pointers to the volumes, etc.
Let's discuss this please.

**Comment by AdrianSoundy** [comment]
Sorry a meant to leave some notes for discussion for this PR before review. That why I hadn't requested review yet :-)

When started testing the SD card stuff I found the storagemanager RemovableDeviceInsertion event was being used to indicate when a removable device had been mounted in file system although I thought it was for the card detect. The event logic for card detect was no longer working but looked like it had been calling the same event which didn't make sense.  I fixed event so card detect called RemovableDeviceInsertion event then found the args for event didn't make sense. It was passing a drive path for I: drive for a card which wasn't mounted yet. I hope this makes sense.

So that's why I decided to create a new event for card detect with new args and leave the mount & unmount events alone.  

Probably the RemovableDeviceInsertion event is the wrong name as it is being used for when a device is mounted. Should we rename it ?  And call the Card Detect event that name ?

The card detect event I routed to the relevant sdcard object which I thought made more sense.

Everything is working now but I do think maybe some renaming will be in order. What do you think.

@josesimoes

**Comment by AdrianSoundy** [comment]
The storage list is for mounted drives. With the card detect it isn't mounted yet so could use that for lookup.

File: System.IO.FileSystem/System.IO.FileSystem.nfproj (hunk 1 above)
Comment (AdrianSoundy) [reply System.IO.FileSystem/System.IO.FileSystem.nfproj:85]: I had problems updating to latest nugets so removed and added again and that's what i ended with. Seems to build ok.

Comment (AdrianSoundy) [reply System.IO.FileSystem/System.IO.FileSystem.nfproj:85]: Had to manually update version for gitversion in project for it to update properly.

File: System.IO.FileSystem/nanoFramework/SDCardCDParameters.cs (hunk 2 above)
Comment (AdrianSoundy) [reply System.IO.FileSystem/nanoFramework/SDCardCDParameters.cs]: Done

**Comment by AdrianSoundy** [comment]
Unit tests failing because references wrong firmware checksum for filesystem. Won't be fixed until interpreter updated.

**Comment by AdrianSoundy** [comment]
Need to merge this now as matching nanoclr already merged.

**Pull request merged**